scripts/fast_lint.py

```diff
@@ -0,0 +1,117 @@
+
+    return [item.b_path for item in repo.index.diff(common_ancestor)]
+
+
+def run_cmd(cmd: list[str], files: list[str] | bool) -> bool:
+    start = time.time()
+    if not files:
+        logging.info(f"SKIP: `{' '.join(cmd)}`")
+        return True
+
+    if isinstance(files, bool):
+        files = []
+    proc = subprocess.run(cmd + files, text=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    if proc.returncode == 0:
+        logging.info(f"PASS: `{' '.join(cmd)} {'<FILES>' if files else ''}` in {time.time() - start:.2f}s")
+        if proc.stdout:
+            logging.debug(f"stdout: {proc.stdout}")
+    else:
+        logging.info(f"FAIL: `{' '.join(cmd)} {'<FILES>' if files else ''}` in {time.time() - start:.2f}s")
+        if proc.stdout:
+            logging.info(f"stdout: {proc.stdout}")
+
+    return proc.returncode == 0
+
+
+def filter_ext(files: list[str], ext: list[str]) -> list[str]:
+    return [f for f in files if any(f.endswith(e) for e in ext)]
+
+
+def filter_cpp(files: list[str]) -> list[str]:
+    return [f for f in files if f.endswith(".h") or f.endswith(".hpp") or f.endswith(".c") or f.endswith(".cpp")]
+
+
+def main() -> None:
+    start = time.time()
+    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument(
+        "--log-level",
+        dest="log_level",
+        default="INFO",
+        choices=["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"],
+        help="Set the logging level (default: INFO)",
+    )
+    parser.add_argument("--num-threads", type=int, default=8, help="Number of threads to use (default: 4)")
+    parser.add_argument(
+        "files",
+        metavar="file",
+        type=str,
+        nargs="*",
+        help="File paths. Empty = all files, recursively.",
+    )
+
+    args = parser.parse_args()
+    logging.basicConfig(level=args.log_level)
+
+    script_dirpath = os.path.dirname(os.path.realpath(__file__))
+    root_dirpath = os.path.abspath(f"{script_dirpath}/..")
+    os.chdir(root_dirpath)
+
+    if args.files:
+        files = args.files
+    else:
+        files = changed_files()
+
+    logging.debug("Checking:")
+    for f in files:
+        logging.debug(f"  {f}")
+
+    jobs = [
+        (["pixi", "run", "codegen", "--check"], True),
+        (["rustfmt", "--check"], filter_ext(files, [".rs"])),
+        (["python", "scripts/lint.py"], files),
+        (["typos"], files),
+        (["taplo", "fmt", "--check"], filter_ext(files, [".toml"])),
+        (["pixi", "run", "clang-format", "--dry-run"], filter_ext(files, [".cpp", ".c", ".h", ".hpp"])),
+        (["ruff", "check", "--config", "rerun_py/pyproject.toml"], filter_ext(files, [".py"])),
+        (["black", "--check", "--config", "rerun_py/pyproject.toml"], filter_ext(files, [".py"])),
+        (["blackdoc", "--check"], filter_ext(files, [".py"])),
+        (["mypy", "--no-warn-unused-ignore"], filter_ext(files, [".py"])),
+    ]
```

Comment:
Out of this PR scope, but it would be nice if all of these would actually be (just_command, file_list) tuple. `justfile` as single source of truth on how to run things...

Reply:
I'm not totally sold on making these all dispatch through just since I'm not sure running them individually always makes sense. However, I've added them all to pixi and updated the pixi deps so they should now run deterministically from the pixi environment.